### PR TITLE
Add serverless validation using lsql library

### PR DESCRIPTION
**Pull request by bishwajit-db**
Add serverless validation using lsql library

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -13,9 +13,10 @@ classifiers = [
   "Programming Language :: Python :: Implementation :: CPython",
 ]
 dependencies = [
-  "databricks-sdk>=0.18,<0.22",
+  "databricks-sdk>=0.18.0,<=0.22.0",
   "sqlglot==22.4.0",
-  "databricks-labs-blueprint[yaml]>=0.2.3,<0.5.0"
+  "databricks-labs-blueprint[yaml]>=0.2.3,<0.5.0",
+  "databricks-labs-lsql==0.2.2",
```
Comment (nfx) [inline pyproject.toml:19]:
```suggestion
  "databricks-labs-lsql>=0.2.2",
```
To avoid conflicts

File: pyproject.toml [hunk 2]
```diff
@@ -13,9 +13,10 @@ classifiers = [
   "Programming Language :: Python :: Implementation :: CPython",
 ]
 dependencies = [
-  "databricks-sdk>=0.18,<0.22",
+  "databricks-sdk>=0.18.0,<=0.22.0",
```
Comment (nfx) [inline pyproject.toml:16]:
```suggestion
  "databricks-sdk>=0.18.0",
```

File: pyproject.toml [hunk 3]
```diff
@@ -13,9 +13,10 @@ classifiers = [
...
   "sqlglot==22.4.0",
-  "databricks-labs-blueprint[yaml]>=0.2.3,<0.5.0"
+  "databricks-labs-blueprint[yaml]>=0.2.3,<0.5.0",
```
Comment (nfx) [inline pyproject.toml:18]:
```suggestion
  "databricks-labs-blueprint[yaml]>=0.2.3",
```

File: src/databricks/labs/remorph/helpers/validate.py [hunk 4]
```diff
@@ -71,46 +62,59 @@ def validate_format_result(self, config: MorphConfig, input_sql: str):
...
-
-            if "Hive support is required to CREATE Hive TABLE (AS SELECT).;" in str(aex):
-                logger.debug(f"Analysis Exception : IGNORED: {str(aex)}")
-                return True, str(aex)
-
-            logger.debug(f"Unknown Exception: {str(aex)}")
-            return False, str(aex)
+        except DatabricksError as dbe:
+            err_msg = str(dbe)
+            if "[PARSE_SYNTAX_ERROR]" in err_msg:
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/validate.py:84]: Most of these exceptions are already concrete types in LSQL, refactor to multiple except clauses. Add more mapping to lsql if necessary

File: src/databricks/labs/remorph/transpiler/execute.py [hunk 5]
```diff
@@ -48,7 +50,7 @@ def process_file(config: MorphConfig, input_file: str | Path, output_file: str |
                     w.write(output)
                     w.write("\n;\n")
                 else:
-                    validate = Validate(config.sdk_config)
+                    validate = Validate(WorkspaceClient(config=config.sdk_config))
```
Comment (nfx) [inline src/databricks/labs/remorph/transpiler/execute.py:53]: Inject workspace client through the function argument, it's available in the cli. Every workspace client must have product name and product version for tracking

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/cli.py [hunk 6]
```diff
@@ -53,7 +53,7 @@ def transpile(
         sdk_config=w.config,
     )
 
-    status = morph(config)
+    status = morph(config, w)
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/cli.py:56]: Can you add a readme document on how to create this workspace client , if done externally with required fields.

**Comment by sundarshankar89** [comment]
@bishwajit-db  I fixed the make fmt issue, linter checks please fix those

File: src/databricks/labs/remorph/helpers/validation.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        Returns:
+        - tuple: A tuple containing a boolean indicating whether the query is valid or not,
+        and a string containing a success message or an exception message.
+        """
+        # When variables is mentioned Explain fails we need way to replace them before explain is executed.
+        explain_query = f'EXPLAIN {query.replace("${", "`{").replace("}", "}`").replace("``", "`")}'
+        try:
+            rows = list(sql_backend.fetch(explain_query))
+            if not rows:
+                return False, "No results returned from explain query."
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/helpers/validation.py:73]: do we need this check given explain will always return rows or error out.

Comment (bishwajit-db) [reply src/databricks/labs/remorph/helpers/validation.py:73]: It is just for safeguarding against erroneous result.

**Review by sundarshankar89 — APPROVED** [review]
LGTM

**Review by nfx — APPROVED** [review]
Lgtm

**Pull request merged**